songpal/device.py

```diff
@@ -264,11 +329,110 @@ async def activate_system_update(self) -> None:
 
     async def get_inputs(self) -> List[Input]:
         """Return list of available outputs."""
-        res = await self.services["avContent"]["getCurrentExternalTerminalsStatus"]()
+        if "avContent" in self.services:
+            res = await self.services["avContent"][
+                "getCurrentExternalTerminalsStatus"
+            ]()
+            return [
+                Input.make(services=self.services, **x)
+                for x in res
+                if "meta:zone:output" not in x["meta"]
+            ]
+        else:
+            if self._upnp_discovery is None:
+                raise SongpalException(
+                    "avContent service not available and UPnP fallback failed"
+                )
+
+            return await self._get_inputs_upnp()
+
+    async def _get_upnp_services(self):
+        requester = AiohttpRequester()
+        factory = UpnpFactory(requester)
+
+        if self._upnp_device is None:
+            self._upnp_device = await factory.async_create_device(
+                self._upnp_discovery.upnp_location
+            )
+
+        if self._upnp_renderer is None:
+            media_renderers = await DmrDevice.async_search(timeout=1)
```

Comment:
One possibility to skip multiple discovery queries would be to modify the original `Discover.discover()` not to filter with the scalar web api ST, and do filtering during the discovery process.